File: leetcode/class_design/lc_155_min_stack.py
'''
https://leetcode.com/problems/min-stack/description/
'''
class MinStack:

    def __init__(self):
        # init our regular stack
        self.stack = []
        # init our min_stack
        # the top of this stack will
        # always contain the smallest value
        # because we will insert only 
        # smaller values than the top into it
        self.min_stack = []

    def push(self, val: int) -> None:
        # add the value to our regular stack
        self.stack.append(val)

        # if we have an empty min stack
        # or if the new value is smaller
        # than the top of the min stack, then append
        # to the min stack
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)
        # else the new value > top of min stack
        else:
            # then get the current top value
            top = self.min_stack[-1]
            # and readd the top value
            # to maintain len(min_stack) = len(stack)
            # and the property that the top
            # of the min stack is always the smallest value
            self.min_stack.append(top)

    def pop(self) -> None:
        # remove the top element from main stack
        self.stack.pop()
        # remove the top element from the min stack
        self.min_stack.pop()

    def top(self) -> int:
        # return the top of the main stack
        return self.stack[-1]

    def getMin(self) -> int:
        # return the top of the min stack
        return self.min_stack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: leetcode/cpp/class_design/lc_155_min_stack.cpp
#include <stack>

/**
 * https://leetcode.com/problems/min-stack/description/
 *
 * Time complexity: O(1) for all functions
 * Space complexity: O(n) where n = number of variables inserted into MinStack
 */
class MinStack {
public:
    // first define our stacks
    // the first is a regular stack
    std::stack<int> regularStack;
    // the second is a monotonic decreasing stack
    // so the top element is the smallest value out
    // of all elements defined in the regularStack
    std::stack<int> minStack;

    MinStack() {

    }

    void push(int val) {
        // we can directly push onto the regular stack
        regularStack.push(val);

        // if we don't have any values in our min stack
        // or the value we are adding is smaller than the
        // current smallest value, then we will also directly
        // push onto the min stack
        if (minStack.size() == 0 || val < minStack.top()) {
            minStack.push(val);
        // otherwise, we will continue to use the previous min value
        // and push that onto the min stack
        } else {
            minStack.push(minStack.top());
        }
    }

    // for both regular and min stacks, we can pop from them
    // normally, as long as they have an element in them
    void pop() {
        if (regularStack.size() > 0) {
            regularStack.pop();
        }
        if (minStack.size() > 0) {
            minStack.pop();
        }
    }

    // we can simply return the top value of ther regular stack
    int top() {
        return regularStack.top();
    }

    // we can simply return the top value of the min stack
    int getMin() {
        return minStack.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
